clients/unshaded/src/main/java/tachyon/client/TachyonFile.java

```diff
@@ -198,17 +202,24 @@ public String getLocalFilename(int blockIndex) throws IOException {
    * @throws IOException if the underlying file does not exist or its metadata is corrupted
    */
   public List<String> getLocationHosts() throws IOException {
-    List<String> ret = new ArrayList<String>();
+    Set<String> ret = Sets.newHashSet();
     if (getNumberOfBlocks() > 0) {
-      List<NetAddress> locations = getClientBlockInfo(0).getLocations();
-      if (locations != null) {
-        for (NetAddress location : locations) {
+      // under FS locations
+      List<NetAddress> underFsLocations = getClientBlockInfo(0).getUnderFsLocations();
+      if (underFsLocations != null) {
+        for (NetAddress location : underFsLocations) {
           ret.add(location.host);
         }
       }
+      List<BlockLocation> blockLocations = getClientBlockInfo(0).getBlockInfo().getLocations();
```

Comment:
The old code would usually sort the tachyon locations before the ufs locations, with the assumption that the first locations were "faster" than the later locations. Therefore, the tachyon locations should be added first, and the ufs second. Also, this would mean the set would have to be a list.